### PR TITLE
guard against negative and trivial potentials

### PR DESCRIPTION
**Pull request by totto82**
Negative and trivial potentials creates problems for the guiderates.

### CODE REVIEW

File: opm/simulators/wells/GasLiftSingleWell_impl.hpp [hunk 1]
```diff
@@ -112,6 +112,11 @@ computeWellRates_(
             -potentials[this->water_pos_]);
         displayDebugMessage_(msg);
     }
+
+    for (size_t phase = 0; phase < potentials.size(); ++phase){
+        // make sure the potentials are positive
```
Comment (bska) [inline opm/simulators/wells/GasLiftSingleWell_impl.hpp:117]: This comment disagrees with the code so either
1. The comment is wrong
2. The code is wrong
3. Both the code and the comment are wrong

File: opm/simulators/wells/MultisegmentWell_impl.hpp [hunk 2]
```diff
@@ -285,8 +285,9 @@ namespace Opm
             // if the rates are trivial we are most probably looking at the newly
             // opened well and we therefore make the affort of computing the potentials anyway.
             if (std::abs(total_rate) > 0) {
+                double sign = this->isInjector() ? 1.0:-1.0;
```
Comment (bska) [inline opm/simulators/wells/MultisegmentWell_impl.hpp:288]: Please make this `const`.

File: opm/simulators/wells/MultisegmentWell_impl.hpp [hunk 3]
```diff
@@ -302,6 +303,11 @@ namespace Opm
         }
         deferred_logger.debug("Cost in iterations of finding well potential for well "
                               + this->name() + ": " + std::to_string(debug_cost_counter_));
+
+        double sign = this->isInjector() ? 1.0:-1.0;
```
Comment (bska) [inline opm/simulators/wells/MultisegmentWell_impl.hpp:307]: Please make this `const`.

File: opm/simulators/wells/StandardWell_impl.hpp [hunk 4]
```diff
@@ -1899,8 +1899,9 @@ namespace Opm
             // if the rates are trivial we are most probably looking at the newly
             // opened well and we therefore make the affort of computing the potentials anyway.
             if (std::abs(total_rate) > 0) {
+                double sign = this->isInjector() ? 1.0:-1.0;
```
Comment (bska) [inline opm/simulators/wells/StandardWell_impl.hpp:1902]: Please make this `const`.

File: opm/simulators/wells/StandardWell_impl.hpp [hunk 5]
```diff
@@ -1917,6 +1918,11 @@ namespace Opm
             // the well has a THP related constraint
             well_potentials = computeWellPotentialWithTHP(ebosSimulator, deferred_logger, well_state);
         }
+
+        double sign = this->isInjector() ? 1.0:-1.0;
```
Comment (bska) [inline opm/simulators/wells/StandardWell_impl.hpp:1922]: Please make this `const`.

**Comment by totto82** [comment]
I already found a case where the epsilon creates problems. I will need to rethink how we handle negative potentials and come back with a better solution.

**Comment by totto82** [comment]
jenkins build this please

**Review by bska — COMMENTED** [review]
Please don't make objects mutable when they don't have to be.

File: opm/simulators/wells/GasLiftSingleWell_impl.hpp (hunk 1 above)
Comment (totto82) [reply opm/simulators/wells/GasLiftSingleWell_impl.hpp:117]: The comment is wrong. It should be negative. The sign is changed later.

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
@bska Do you mind reviewing this after my latest changes.

File: opm/simulators/wells/BlackoilWellModelGeneric.cpp [hunk 6]
```diff
@@ -2374,4 +2374,20 @@ runWellPIScaling(const int timeStepIdx,
...
+guideRateUpdateIsNeeded() const {
+    int update = 0;
+    for (const auto& well : well_container_generic_) {
+        if (well->changedToOpenThisStep()) {
+            update = 1;
+            break;
+        }
+    }
+    update = comm_.max(update);
+    return update;
```
Comment (bska) [inline opm/simulators/wells/BlackoilWellModelGeneric.cpp:2389]: I might possibly consider writing this in terms of [`std::any_of()`](https://en.cppreference.com/w/cpp/algorithm/all_any_none_of)
```C++
const auto need_update =
    std::any_of(this->well_container_generic_.begin(),
                this->well_container_generic_.end(),
    [](const WellInterfaceGeneric* well)
{
    return well->changedToOpenThisStep();
});

return this->comm_.max(static_cast<int>(need_update));
```
This, I think, is more expressive than a raw loop with a nested conditional.

File: opm/simulators/wells/GasLiftSingleWell_impl.hpp [hunk 7]
```diff
@@ -112,6 +112,11 @@ computeWellRates_(
             -potentials[this->water_pos_]);
         displayDebugMessage_(msg);
     }
+
+    for (size_t phase = 0; phase < potentials.size(); ++phase){
+        // make sure the potentials are negative
+        potentials[phase] = std::min(0.0, potentials[phase]);
+    }
```
Comment (bska) [inline opm/simulators/wells/GasLiftSingleWell_impl.hpp:119]: You could write this as
```C++
for (auto& potential : potentials) {
    potential = std::min(0.0, potential);
}
```

File: opm/simulators/wells/WellInterface_impl.hpp [hunk 8]
```diff
@@ -482,7 +482,23 @@ namespace Opm
...
             }
         }
-
+        if (this->operability_status_.has_negative_potentials) {
+            auto well_state_copy = well_state;
+            std::vector<double> potentials;
+            try {
+                computeWellPotentials(ebosSimulator, well_state_copy, potentials, deferred_logger);
+            } catch (const std::exception& e) {
+                const std::string msg = std::string("well ") + this->name() + std::string(": computeWellPotentials() failed during for re-computing: ") + e.what();
```
Comment (bska) [inline opm/simulators/wells/WellInterface_impl.hpp:491]: Tiny nit: This diagnostic message is confusing.  What is
```
failed during for re-computing
```
supposed to say?

**Review by bska — COMMENTED** [review]
Thank you for the updates.  This looks better to me.  I just have a couple of minor remarks about something I'd personally write slightly differently, but otherwise I'd be happy to merge this.

**Comment by totto82** [comment]
Thanks @bska for good suggestions. I have changed the code accordingly.

File: opm/simulators/wells/StandardWell_impl.hpp [hunk 9]
```diff
@@ -1892,15 +1893,17 @@ namespace Opm
...
             for (int phase = 0; phase < np; ++phase){
-                total_rate += ws.surface_rates[phase];
+                total_rate += sign * ws.surface_rates[phase];
             }
             // for pressure controlled wells the well rates are the potentials
             // if the rates are trivial we are most probably looking at the newly
             // opened well and we therefore make the affort of computing the potentials anyway.
-            if (std::abs(total_rate) > 0) {
+            if (total_rate > 0) {
+                const double sign = this->isInjector() ? 1.0:-1.0;
```
Comment (bska) [inline opm/simulators/wells/StandardWell_impl.hpp:1904]: The latest build check just alerted me to the fact that this `sign` object [shadows](https://ci.opm-project.org/job/opm-simulators-PR-builder/3326/gcc/fileName.1498975879/) an object with the same value in the outer scope.  You can probably just delete this object.

**Review by bska — COMMENTED** [review]
Just one tiny issue that I didn't notice until running the build check. There's a redundant/shadowing definition of `sign` in `StandardWell<>::computeWellPotentials()`.  Once we fix that I'll merge this.

**Review by bska — APPROVED** [review]
Thanks a lot for the final updates.  I'll merge this into master now.

**Pull request merged**